staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -292,9 +292,9 @@ func (dm *discoveryManager) fetchFreshDiscoveryForService(gv metav1.GroupVersion
 			lastUpdated: now,
 		}
 
-		// Save the resolve, because it is still useful in case other services
-		// are already marked dirty. THey can use it without making http request
-		dm.setCacheEntryForService(info.service, cached)
```

Comment:
prior to this change, the discovery check interval was effectively once every minute for each apiservice.  After this change it is an unbounded number of checks per unit time.  Can we get back the "once every minute" aspect of the original?

Reply:
The one check every minute is done here: https://github.com/kubernetes/kubernetes/blob/ed98c1da6d5fd1382a97663038c0fedbc02ebf9f/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go#L426-L439 and is untouched in this PR.

The behavior before is that the check is once a minute, but if multiple apiservices share the same service, a fetch will not be performed (and cached result used instead) so the behavior was effectively **at most** one check per apiservice per minute. Now, the behavior is **exactly** one check per minute per apiservice for apiservices that need to use the legacy fallback, while for apiservices who supported aggregated discovery, the behavior is unchanged **at most** one check per apiservice per minute.

Reply:
>Now, the behavior is exactly one check per minute per apiservice for apiservices that need to use the legacy fallback, while for apiservices who supported aggregated discovery

If AddAPIService is called more than once for whatever reason (maybe status got touched?), after this PR it is checked more than once per minute, right?

Reply:
I think also prior to this PR, calling `AddAPIService` more than once per minute would result in the service being added to the queue every time, if it had synced by then, and result in refetching the discovery document.

Before and after this PR, the controller conflates an update to the APIService as a possible change to the underlying service. I think it is worth filing a separate issue to fix.

I think at the very least maybe we should add a check to at least see if the `Spec.Service` has changed at all before marking the APIService/service as dirty.

Reply:
>I think also prior to this PR, calling AddAPIService more than once per minute would result in the service being added to the queue every time

I thought the dirty time was set for one minute later, but I can't find that line now.  I'll double check tomorrow, but if I can't find it, I'm ok with this change.

>I think at the very least maybe we should add a check to at least see if the Spec.Service has changed at all before marking the APIService/service as dirty.

The check needs to be slightly smarter, since you do need to periodically check the content to see if the etag changed in the modern case and in the legacy case, you probably want to check every few minutes.

Reply:
Looks like `now` everywhere.  I must have mis-remembered.